swupd/manifest.go

```diff
@@ -342,72 +342,113 @@ func (m *Manifest) sortFilesVersionName() {
 	})
 }
 
-func (m *Manifest) addDeleted(oldManifest *Manifest) {
-	for _, df := range oldManifest.DeletedFiles {
-		if df.findFileNameInSlice(m.Files) == nil {
-			m.Files = append(m.Files, df)
-			m.DeletedFiles = append(m.DeletedFiles, df)
-		}
-	}
+func (m *Manifest) appendFile(f *File) {
+	m.Files = append(m.Files, f)
 }
 
 // linkPeersAndChange
 // At this point Manifest m should only have the files that were present
 // in the chroot for that manifest. Link delta peers with the oldManifest
-// if the file in the oldManifest is not deleted or ghosted
-func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) int {
-	changed := 0
-	for _, of := range oldManifest.Files {
-		if match := of.findFileNameInSlice(m.Files); match != nil {
+// if the file in the oldManifest is not deleted or ghosted.
+// Expects m and oldManifest files lists to be sorted by name only
+func (m *Manifest) linkPeersAndChange(oldManifest *Manifest) (int, int, int) {
+	var unchanged, changed, removed, added []*File
+	i := 0
+	j := 0
+	mFilesLen := len(m.Files)
+	omFilesLen := len(oldManifest.Files)
+	for i < mFilesLen && j < omFilesLen {
+		nf := m.Files[i]
+		of := oldManifest.Files[j]
+		if nf.Name == of.Name {
+			// if it is the same name check if anything about the file has changed.
+			// if something has changed update the version to current version and
+			// record that the file has changed
 			if of.Status == statusDeleted || of.Status == statusGhosted {
+				// don't worry about ghosted or deleted files
+				i++
+				j++
 				continue
 			}
 
-			match.DeltaPeer = of
-			of.DeltaPeer = match
-			if !sameFile(match, of) {
-				match.Version = m.Header.Version
-				changed++
+			// set up peers
+			nf.DeltaPeer = of
+			of.DeltaPeer = nf
+			if !sameFile(nf, of) {
+				// if the file isn't exactly the same, record the change
+				// and update the version
+				nf.Version = m.Header.Version
+				changed = append(changed, nf)
 			} else {
-				match.Version = of.Version
+				// otherwise the version is the same as the old file
+				nf.Version = of.Version
+				unchanged = append(unchanged, nf)
+			}
+			// advance indices for both file lists since we had a match
+			i++
+			j++
+		} else if nf.Name < of.Name {
+			// if the file does not exist in the old manifest it is a new
+			// file in this manifest. Update the version and record that
+			// it is added.
+			nf.Version = m.Header.Version
+			added = append(added, nf)
+			// look at next file in current manifest
+			i++
+		} else {
+			// if the file exists in the old manifest and does not *yet* exist
+			// in the new manifest, it was deleted. Set up the deleted file
+			// and prepend it to the files list via newDeleted()
+			if of.Status == statusDeleted || of.Status == statusGhosted {
+				j++
+				continue
 			}
+			m.newDeleted(of)
```

Comment:
We could just delay the call to m.newDeleted until after we do the rename stuff, and just do it on the removed slice